### PR TITLE
fix: Improved behavior when loading CJS modules from import statements

### LINKED ISSUE #667: Need improvement on calling as CJS module from ESM
Calling the CJS module from the ESM has been improved by the resolution of #641, but we need further improvements.

See also.
https://github.com/awslabs/llrt/issues/641#issuecomment-2450988284

In addition, the following PRs would be helpful as a path that should not be taken.
https://github.com/awslabs/llrt/pull/664#issuecomment-2453113856

--
This is by design. We need to be able to distinguish cjs from regular mjs files. I’ve considered options for trying to automatically detect this or simply treat all require imports as CJS. What’s your take?

_Originally posted by @richarddavison in https://github.com/awslabs/llrt/issues/641#issuecomment-2450550500_

### PR DESCRIPTION
**Pull request by nabetti1720**
### Issue # (if available)

Fixed #667
Related #641 

### Description of changes

Fixed an issue where the error `SyntaxError: Could not find export 'default' from '...'` occurred when loading a CJS module from an import statement.

- When resolving a Node.js Package, we now also determine whether the package is CJS.
- After that, if we determine that it is a CJS module called by import statement, we now use prefix `__cjsm:` to achieve this behavior.

The following use case, which was an issue in #641, will now work:

```javascript
// powertoolsLogger.js
import { Logger } from '@aws-lambda-powertools/logger'
const logger = new Logger()
logger.info('Hello world')
```

```
% llrt powertoolsLogger.js 
{"level":"INFO","message":"Hello world","sampling_rate":0,"service":"service_undefined","timestamp":"2024-11-17T01:52:13.459Z"}
```

### Checklist

- [x] Created unit tests in `tests/unit` and/or in Rust for my feature if needed
- [x] Ran `make fix` to format JS and apply Clippy auto fixes
- [x] Made sure my code didn't add any additional warnings: `make check`
- [x] Added relevant type info in `types/` directory
- [x] Updated documentation if needed ([API.md](API.md)/[README.md](README.md)/Other)

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: llrt_core/src/module_loader/resolver.rs [hunk 1]
```diff
@@ -548,7 +553,7 @@ fn load_package_self(ctx: &Ctx<'_>, x: &str, dir: &str, is_esm: bool) -> Result<
     //    "." + X.slice("name".length), `package.json` "exports", ["node", "require"])
     //    <a href="esm.md#resolver-algorithm-specification">defined in the ESM resolver</a>.
     // 6. RESOLVE_ESM_MATCH(MATCH)
-    if let Ok(path) = package_exports_resolve(&package_json, &name, is_esm) {
+    if let Ok((path, _)) = package_exports_resolve(&package_json, &name, is_esm) {
```
Comment (nabetti1720) [inline llrt_core/src/module_loader/resolver.rs:556]: We don't yet know if we need the same implementation here. Even if we did, the path is still relative, so adding a prefix at this point would be a mistake.

**Comment by nabetti1720** [comment]
It seems that a similar issue is now being reported in QuickJS-NG (numerous warnings occur when building in Visual Studio 2022)...

image

**Comment by richarddavison** [comment]
I think it would be valuable to add some tests for require to handle edge cases. We can create a mock `node_modules` folder in `fixtures` and add fake packages in there (not real packages)

**Comment by nabetti1720** [comment]
> I think it would be valuable to add some tests for require to handle edge cases. We can create a mock `node_modules` folder in `fixtures` and add fake packages in there (not real packages)

I've written more specific test code for some of the require patterns that have been supported so far. This PR covers the patterns that can confirm this use case and the patterns where regressions have recently occurred.

**Review by richarddavison — APPROVED** [review]
